test-data/unit/check-overloading.test

```diff
@@ -3717,3 +3715,39 @@ def relpath(path: str) -> str: ...
 @overload
 def relpath(path: unicode) -> unicode: ...
 [out]
+
+[case testUnionMathTrickyOverload1]
+from typing import Union, overload
+
+@overload
+def f(x: int, y: int) -> int: ...
+@overload
+def f(x: object, y: str) -> str: ...
+def f(x):
+    pass
+
+x: Union[int, str]
+y: Union[int, str]
+f(x, y)
+[out]
+main:12: error: Argument 1 to "f" has incompatible type "Union[int, str]"; expected "int"
+main:12: error: Argument 2 to "f" has incompatible type "Union[int, str]"; expected "int"
+
+[case testUnionMathTrickyOverload2]
+from typing import overload, Union, Any
+
+class C:
+    def f(self, other: C) -> C: ...
+
+class D(C):
+    @overload  # type: ignore
+    def f(self, other: D) -> D: ...
+    @overload
+    def f(self, other: C) -> C: ...
+    def f(self, other): ...
+
+x: D
+y: Union[D, Any]
+# TODO: update after we decide on https://github.com/python/mypy/pull/5254
+reveal_type(x.f(y))  # E: Revealed type is 'Any'
```

Comment:
Just a comment to myself, we need to fix this before merging this PR, Jukka prefers `Union[D, Any]` here, which actually the result a unioned match, but direct match is `Any`, so:
* #5254 should go in first
* after that we might tweak the choice between direct and unioned match